### ISSUE TITLE
Copy task should copy files only if changed

### ISSUE THREAD

**Issue by kno10** [gradle/gradle#1643]
I am using some big dependency jars (CoreNLP language models).
It slows down my build a lot that they are copied every time.

Instead, the copy task should skip copying if file date and file size match.

An earlier report of the same, which suggests a few different strategies:
https://issues.gradle.org/browse/GRADLE-1264

Ant does it this way: "By default, files are only copied if the source file is newer than the destination file, or when the destination file does not exist."
https://ant.apache.org/manual/Tasks/copy.html

**Comment by bmuschko** [comment]
They should not be copied every time. Please provide a concrete code example that reproduces the issue.

**Comment by kno10** [comment]
What I am basically using is this:
```
task collectDependencies(type: Copy) {
	from configurations.runtime
	into "$project.buildDir/libs"
}
```
but with a modification to preserve timestamps and an additional from. If I recall correctly, this basic version already exhibited the same problem, copying files even when not modified (and because of this, updating the file date).

Can you show me where Gradle checks the timestamp? I could not find such code.

**Comment by big-guy** [comment]
@bmuschko I think @kno10 is saying that _when_ the copy task needs to execute, it copies all the files, even if some of the files are unchanged.  That makes sense based on how the `Copy` task works--it doesn't try to incrementally copy files. This is actually a little harder to do since we have ways of modifying a file while copying it, so it's not enough that the source and destination have the same date/size.  We also need to make sure the configuration of the `Copy` task is the same.

@kno10 What is the task used to do, e.g. build a distribution?

**Comment by kno10** [comment]
I use this to collect all dependencies. They will then be copied to the workers.

Indeed: *some* of the 'dependencies' (the jars with my own code) will have changed. I don't want to copy all the dependencies again, because some of them are huge - the CoreNLP language models are over one GB.

Do I understand this right: if I invert the logic, such that for every dependency there is a separate copy task, then it would work as desired?

**Comment by nicolas-albert** [comment]
Hi,
Try to use "Sync" instead of "Copy".

**Comment by kno10** [comment]
For all I can tell, Sync will likely suffer from the same problem, and will re-write all files on every run, rather than checking files for modification, and overwriting modified files only.
The main difference between Sync and Copy appears to be that Sync will remove other contents of the destination folder.

**Comment by bungrudi** [comment]
I am also very interested to see this incremental feature implemented in Copy and Sync task.
For now I have to resort ant.copy.

**Comment by JohannBlake** [comment]
Was this feature ever added? It's been a year.

**Comment by rsivan** [comment]
I want to add, please make sure that when this feature is added there will be a way to specify a 'force' copy (override regardless of timestamp).

**Comment by vonkendu** [comment]
Are there any news? Copy tasks still behaves this way. It's been quite a while now and it does force many to use ant.copy which is quite awkward.

**Comment by mikejhill** [comment]
Was running into this issue today. I tried manually comparing the timestamps as mentioned [here](https://discuss.gradle.org/t/using-war-copyspec-for-ant-copy-task/23711/4), but it looks like Gradle doesn't retain the last-modified timestamp when copying files (#1252). I was able to use kno10's workaround ([here](https://github.com/gradle/gradle/issues/1252#issuecomment-421744595)) for this, and then after that manually comparing the timestamps worked great. In our specific case, the resulting execution time is much closer to that of `ant.copy` for incremental copies.

```kotlin
// Reduce cost by only copying files which have been modified
eachFile {
    val targetFile = File(destinationDir, path)
    if (file.lastModified() == targetFile.lastModified() && file.length() == targetFile.length()) {
        exclude()
    }
}

// ... other actions ...

// Retain timestamps from source files (see https://github.com/gradle/gradle/issues/1252#issuecomment-421744595)
// This should take place after all exclusion filters in order to prevent unnessary timestamp updates to unmodified files
val copyDetails = mutableListOf<FileCopyDetails>()
eachFile { copyDetails += this }
doLast {
    copyDetails.forEach { details ->
        File(destinationDir, details.path).takeIf { it.exists() }?.setLastModified(details.lastModified)
    }
}
```

**Comment by kno10** [comment]
@stale is an extremely stupid bot.

**Comment by tlf30** [comment]
I agree. First, having a bot marking things as stale just causes issues that are relevant to be lost because people have nothing else to say about the issue until it is fixed.

Second, for projects attempting to handle large numbers of files, the ability to only copy what has changed would be very beneficial as it would speed up builds, in my case by several minutes.

**Comment by JD10NN3** [comment]
Is it implemented?

**Comment by kno10** [comment]
No, AFAIK you still need to do a manual workaround yourself.
But the stupid bot auto-closes issues, and there is no "reopen" button, apparently you have to ping @bmuskalla

**Comment by tlf30** [comment]
@wolfs why was this closed as not planned?

**Comment by wolfs** [comment]
> @wolfs why was this closed as not planned?

It already was closed. Thanks for pinging so I can re-open it. Seems like the issue is still interesting for folks.

**Comment by tlf30** [comment]
Great, thank you!

**Comment by CLOVIS-AI** [comment]
This feature would be useful for integration with other built tools, e.g. Vite.

The Kotlin plugin even has its task for this: https://github.com/JetBrains/kotlin/blob/032e1cdbbc89268423302fad3d605933dcb7e5c0/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/targets/js/ir/DefaultIncrementalSyncTask.kt#L17
